design/general-progress-monitoring.md

```diff
@@ -0,0 +1,462 @@
+
+Velero currently has backup/restore phases "InProgress" and "Completed".  A backup moves to the
+Completed phase when all of the volume snapshots have completed and the Kubernetes metadata has been
+written into the object store.  However, the actual data movement may be happening in the background
+after the backup has been marked "Completed".  The backup is not actually a stable backup until the
+data has been persisted properly.  In some cases (e.g. AWS) the backup cannot be restored from until
+the snapshots have been persisted.
+
+Once the snapshots have been taken, however, it is possible for additional backups or restores (as
+long as they don't use not-yet-completed backups) to be made without interference.  Waiting until
+all data has been moved before starting the next backup will slow the progress of the system without
+adding any actual benefit to the user.
+
+A new backup/restore phase, "WaitingForPluginOperations" will be introduced.  When a backup or
+restore has entered this phase, Velero is free to start another backup/restore.  The backup/restore
+will remain in the "WaitingForPluginOperations" phase until all BIA/RIA operations have completed
+(for example, for a volume snapshotter, until all data has been successfully moved to persistent
+storage).  The backup/restore will not fail once it reaches this phase.  If the backup is deleted
+(cancelled), the plug-ins will attempt to delete the snapshots and stop the data movement - this may
+not be possible with all storage systems.
+
+### State progression
+
+(Note that the diagrams have not yet been updated to reflect the terminology changes associated with
+the revised design.)
+
+![image](UploadFSM.png)
+### New
+When a backup/restore request is initially created, it is in the "New" phase.  
+
+The next state is either "InProgress" or "FailedValidation"
+
+### FailedValidation
+If the backup/restore request is incorrectly formed, it goes to the "FailedValidation" phase and
+terminates
+
+### InProgress
+When work on the backup/restore begins, it moves to the "InProgress" phase.  It remains in the
+"InProgress" phase until all pre/post execution hooks have been executed, all snapshots have been
+taken and the Kubernetes metadata and backup/restore info is safely written to the object store
+plug-in.
+
+In the current implementation, Restic backups will move data during the "InProgress" phase.  In the
+future, it may be possible to combine a snapshot with a Restic (or equivalent) backup which would
+allow for data movement to be handled in the "WaitingForPluginOperations" phase,
+
+The next phase is either "Completed", "WaitingForPluginOperations", "Failed" or "PartiallyFailed".
+Backups/restores which would have a final phase of "Completed" or "PartiallyFailed" may move to the
+"WaitingForPluginOperations" or "WaitingForPluginOperationsPartialFailure" state.  A backup/restore
+which will be marked "Failed" will go directly to the "Failed" phase.  Uploads may continue in the
```

Comment:
I prefer not to start uploading before all snapshots are completed, because many providers don't have a API to stop in-progress snapshotting, and if backup or restore failed, new backup and restore can start. It's possible that new backup and restore also require to snapshot or restore the the same disk. If there is a in-progress operation of the disk, I think there's high possibility that error would be raised.

Reply:
@Lyndon-Li Regarding those cases you mention, I suspect most of those would be individual implementation details for specific plugins, rather than part of the snapshot API. But to briefly go through the four use cases:
- What to do with local snapshot data is something that the current volume snapshotter plugins deal with. The API addition here is just to allow monitoring of the upload (which is also already happening in the current plugin). "Provide for users to decide not to keep specific snapshots" doesn't really seem like something we'd want to add to the plugin interface (or if we do, perhaps as a separate later enhancement).
- For volume snapshotters, upload is an integral part of the process - this enhancement just helps to track the (already existing) upload process. For Backup/RestoreItemActions, if there is a specific plugin that wants to perform asynchronous operations conditionally, that is provided for already in this design -- the operation ID returned by the Backup/RestoreItemAction is optional, and if a plugin does not return one, then there is no operation/upload action to monitor.
- Regarding retention policies, again if a plugin wants to alter behavior based on those, it can, but I don't really see a place in the API where we'd want to deal with it -- if we find something to add, it can be done as a future enhancement.
- I'm not sure I see the use case for "don't actually upload the snapshot now" -- we don't offer that for backups as a whole -- a backup is not marked complete until the entire backup is uploaded to object storage (that way if the cluster goes down/crashes/etc. post-backup-completion, it can be recovered right then). The goal of this enhancement is to provide the same for snapshots and other long-running operations initiated from within a plugin action.

Questions:
- The first question is answered above for the four use cases. From Velero's point of view, the upload process is simply something that a plugin initiates, and Velero's role is limited to asking the plugin periodically whether the upload or other operation is now complete. Velero isn't supposed to be able to understand AWS-specific aspects of snapshot upload, or OpenShift-specific aspects of an OpenShift image plugin that copies internal images into or out of an external backup store. It's why we have plugins rather than building all of these operations into velero core.
- The way we make them generic (as described above) is that the operation itself isn't part of the interface. The BIA/RIA Execute() method call has an optional operationID return parameter. If there is no ongoing operation, it's empty. All velero needs to know is to call the plugin again to check on status and pass back the ID. As for configuration, note that the existing BIA/RIA plugins don't involve plugin-specific configuration in the API. That's already how plugins work. We're just extending the plugin interface to allow ongoing operations. So continuing not to support non-generic plugin-specific configurations in the interface is consistent with the current API. If a plugin needs additional information, it is free to document that users should pass this information on via annotations on the resources being backed up or on the velero Backup/Restore CRs, as is currently done by some of the Openshift-specific plugins included in OADP. If we want to add the notion of plugin configuration to the plugin API, that should be a separate future enhancement if there is a need for it.

Reply:
@sseago Yes, I agree that Velero should not understand details of the specific plugins and the cases mentioned above should be implemented by the specific plugins.

As another fact, we need to let the specific plugin know some information so that the plugin could do the actions. If we regard this information as some "configuration" as well, my above questions come down to: how do we share these configuration between Velero and the plugins if the BIA/RIA execution is totally transparent to Velero?
Specifically, The configuration may not be simply user options that could be fulfilled by annotations. For example, whether or when to start an upload/whether to delete the local snapshot after upload succeeds is decided by a combination of below factors:

1. Users explicitly specify how many local snapshot they want to keep. If the current snapshot count is within that number, the local snapshot created by BIA should not be deleted
2. Users explicitly specify how much time later they want to start the upload after snapshot is taken/backup is created
3. Users specify they just want to upload some of the snapshots, not all. For example, the daily/weekly snapshots. Then Velero needs to decide this dynamically according to its retention policy
4. Considering the time of the backup runs, Velero thinks it is not a suitable time to run the upload. For example, the production environment is busy at most of the day time, if Velero runs the upload, it brings a lot burden to the disk, network and may impact the business. Therefore, there is usually a backup window (i.e. from mid night to the next morning), if the upload could not be finished in the window, Velero needs to have a policy (give up the upload or schedule to another time??). This means, Velero needs to decide the time to start the upload according to its schedule dynamically

As you can see above, if 1 and 2 could be fulfilled by a static annotation, the later two ones is decided by Velero dynamically. Then how do Velero and the plugin share this information.

Indeed, the Velero part of the upload policy should not be covered in this design, merely, I think we do need to clarify how do Velero and plugins exchange information/configuration of the upload policies.

Reply:
For sure, as another approach, we can have all these judgements and policy implementation in the specific plugins, then Velero only needs to share the basic & generic things, i.e., retention policies, schedule policies, etc.
The problem with the approach is, there will be duplicated logics in many plugins to implement the policies, and on the other hand, it may be a better way to implement this in Velero centrally.

I am not oppose to this second approach, I am just thinking we should decide & clarify which approach we use and the pros/cons in this design.

Reply:
Indeed, this configuration things will add significant complications. I don't think we should cover any of the details in this design either. Merely, I think when we want to do this some time later, for example, BIA/RIA v3, there is no significant breaking changes.
The primary benefits of the plugin mechanism is we could "close" the dependent logics in the plugin; on the other hand, if we make it too closed (i.e., we expose an "execute" method only in this design), no information & control could be delivered.
Therefore, in this design, we should clarify how we can avoid the breaking changes brought by the "over-closure". Otherwise, we should clarify we accept the "over-closure" and clarify how to support the complicated cases under this "over-closure" situation.

Personally, I think the latter option will touch too many unnecessary/undetermined details, so I prefer the first option (avoid "over-closure"). I make a draft proposal as below:

- We definitely could use the plugin mechanism for the upload/async operations
- Meanwhile, we expose the control of upload/async operations to the caller
- The "execute" method works on any synchronized tasks and return the control of async tasks to the caller if any
- The caller decides when and how to start the async tasks, just like it deciding when to stop them with the Cancel mechanism
- We only expose the control to the caller in any way, so it won't add too many efforts
- We might sacrifice some versatility (so how to make the async operations purely generic might be a question), I am not 100% sure of this for now, but if any, it worths anyway

Reply:
Lets not go there in terms of attempting to give Velero control of the async actions. That adds unnecessary complications that 1) we don't need for any of the 1.10 use cases and 2) adds more places for things to break down. We want to start the  async actions *right now* when the item is backed up and/or restored. If we wait for later, we make the whole process much more fragile. I think the "pass control to velero" idea is something that should be proposed on its own in a future development cycle, since it adds a bunch of complexity that is unnecessary for all of the current use cases.